### ISSUE TITLE
Need an easier way to use dynamically determined symbols

### ISSUE THREAD

**Issue by renkun-ken** [Rdatatable/data.table#2589]
Suppose I have a `data.table` created as follows:

```r
dt <- data.table(x1 = 1:10, x2 = 10:1, x3 = 1:10)
```

I need to do calculations with dynamically determined symbols within `j` like

```
s1 <- "x2"
s2 <- "x3"
```

Two approaches can do the work:

```r
> dt[, get(s1) * get(s2)]
 [1] 10 18 24 28 30 30 28 24 18 10
> dt[, .SD[[1]] * .SD[[2]], .SDcols = c(s1, s2)]
 [1] 10 18 24 28 30 30 28 24 18 10
```

But if the data is very big and `by=` is used, the performance can significantly decay. Also the first approach using `get()` has scoping problem if `s1` or `s2` are themselves columns of `dt`.

Is there any possibility that makes it easier to use dynamically determined symbol without such significant performance decay and scoping problem?

For example, something like

```r
dt[, ..s1 * ..s2]
```

which is inspired by the `..x` notation introduced lately.

**Comment by tdeenes** [comment]
eval(as.name(s1)) is more efficient. But still, it would be great to generalize the .. notation so that it could replace the eval(as.name()) workaround.

**Comment by franknarf1** [comment]
I think the `..` prefix would make sense, maybe a dupe of #633

**Comment by jangorecki** [comment]
AFAIK using language objects is most efficient way, and kind of base R way of handling the task. Two examples:
https://stackoverflow.com/a/37408321
https://stackoverflow.com/a/37008966

**Comment by jangorecki** [comment]
to be addressed by #4304 
```r
dt[, s1 * s2, env=list(s1=s1,s2=s2), verbose=TRUE]
#Argument 'j'  after substitute: x2 * x3
#Detected that j uses these columns: x2,x3 
# [1] 10 18 24 28 30 30 28 24 18 10
```
